### DISCUSSION TITLE
Query Result API

### DISCUSSION THREAD

**Discussion by namedgraph** [RDFLib/rdflib#3013] [Development]
I think it the query [`Result`](https://rdflib.readthedocs.io/en/7.1.1/apidocs/rdflib.html#rdflib.query.Result) API is also awkward af.
> There is a bit of magic here that makes this appear like different Python objects, depending on the type of result.
> If the type is “SELECT”, iterating will yield lists of ResultRow objects
> If the type is “ASK”, iterating will yield a single bool (or bool(result) will return the same bool)
> If the type is “CONSTRUCT” or “DESCRIBE” iterating will yield the triples.

I personally hate this sort of "magic". I would prefer to have less of it, not more, and make things explicit. For example, that the method signature would be
```python
query(query_object ...) -> Union[Graph, Result, bool]
```
where `Graph` is for `DESCRIBE/CONSTRUCT` results, `Result` is for `SELECT` results _only_, and `bool` is for `ASK` results.

Right now having a `Result` object there is no good way (that I know of) to determine what form of query it is coming from.

**Comment by ajnelson-nist** [comment]
This topic is being visited in [Issue 2283](https://github.com/RDFLib/rdflib/issues/2283), and development in progress on [PR 2320](https://github.com/RDFLib/rdflib/pull/2320), though stalled due to availability.

If there are questions about API style preference, my own preference is to defer to definitions or model suggestions from the upstream sources.  I haven't reviewed SPARQL's specification recently enough to recall if there is a `Result` object suggested.  But please, if we can find an upstream source instead of attempting to resolve opinions at only the style layer, let us defer and cite.

**Reply by namedgraph** [reply]
There are [2 forms](https://www.w3.org/TR/sparql11-query/#QueryForms) of SPARQL results: RDF graph/triples (`Graph` in RDFLib) and [SPARQL Query Results]() (with XML, JSON, CSV formats). The latter does not currently have a distinct representation in RDFLib IMO, as all forms of results are conflated into `Result`.

`ASK` result is a special case since it's a single boolean, but it's specified as a case of SPARQL Query Results.
Some engines (e.g. Jena) support RDF dataset/quad results (`Dataset` in RDFLib) as an extension beyond SPARQL 1.1
